firmware/main.c

```diff
@@ -3,32 +3,128 @@
 PortData port_b;
 
+// Indicates whether the SPI Daemon is listening for USB traffic
 volatile bool booted = false;
+// LED Chan: TCC1/WO[0]
+#define PWR_LED_TCC_CHAN 1
+// CC channel 0 on TCC instance 1
+# define PWR_LED_CC_CHAN 0
+// The maximum counter value was chosen to get a 2s period heartbeat
+#define MAX_COUNTER 0x16E30
+// We have 16 slices of a sine wave
+#define NUM_POINT_SLICES 0x10
+// Number of loop iterations in a single slice
+#define COUNTS_IN_SLICE MAX_COUNTER/NUM_POINT_SLICES
+// Evenly spaced points along a sine wave, shifted up by 1, scaled by 0.5
+// multiplied by the max counter value
+const float sin_wave_points[] = {
+    46872,
+    64809.13784181649,
+    80015.50904777585,
+    90176.08144786903,
+    93744,
+    90176.08144786903,
+    80015.50904777585,
+    64809.13784181649,
+    46872.00000000001,
+    28934.862158183518,
+    13728.490952224147,
+    3567.9185521309782,
+    0,
+    3567.9185521309732,
+    13728.490952224138,
+    28934.86215818348
+};
+// The current counter value of our timer
+volatile u32 counter = 0;
+// The index of what point value we have just passed
+volatile u8 current_point_index = 0;
+
+void cancel_breathing() {
+    // Disable the TCC
+    tcc(PWR_LED_TCC_CHAN)->CTRLA.reg = TCC_CTRLA_RESETVALUE;
+    // Disable Boot LED TCC IRQ in the NVIC
+    NVIC_DisableIRQ(TCC1_IRQn);
+    // Set the PWR LED to the default high state
+    pin_out(PIN_LED);
+    pin_high(PIN_LED);
+}
+
+void TCC1_Handler() {
+    // Increment the timer counter
+    counter++;
 
-/*** SysTick ***/
-volatile uint32_t g_msTicks;
-/*** BOOT LED ***/
-unsigned led_next_time = 0;
-
-/* SysTick IRQ handler */
-void SysTick_Handler(void) {
-    g_msTicks++;
-
-    // Boot LED Tasks
-    if (!booted && g_msTicks > led_next_time) {
-        led_next_time += 400;
-        pin_toggle(PORT_A.power);
-        pin_toggle(PORT_B.power);
-        pin_toggle(PIN_LED);
+    // booted is true when the coprocess first gets
+    // a status packet from the spi daemon
+    if (booted == true) {
+        // Stop this breathing animation and cancel interrupts
+        cancel_breathing();
     }
-}
 
-void init_systick() {
-    if (SysTick_Config(48000000 / 1000)) {  /* Setup SysTick Timer for 1 msec interrupts  */
-        while (1) {}                                /* Capture error */
+    // If the counter has hit the max
+    if (counter == MAX_COUNTER) {
+        // Set it back to zero
+        counter = 0;
+        // Set our current point back to the beginning
+        current_point_index = 0;
+    }
+    // Otherwise if this counter has hit the lower boundary of the next point
+    else if (counter == (COUNTS_IN_SLICE * (current_point_index + 1))) {
+        // Set our index to that point
+        current_point_index++;
     }
-    NVIC_SetPriority(SysTick_IRQn, 0x0);
-    g_msTicks = 0;
+
+    // How far along those two points we are
+    u32 proportion = (counter - (current_point_index * COUNTS_IN_SLICE))/COUNTS_IN_SLICE;
```

Comment:
This always truncates to zero, which is why it flickers -- you're just stepping between the 16 points, not performing linear interpolation.